src/modules/launcher/Plugins/Microsoft.Plugin.Program/Storage/PackageRepository.cs

```diff
@@ -0,0 +1,86 @@
+﻿using Microsoft.Plugin.Program.Logger;
+using Microsoft.Plugin.Program.Programs;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Windows.ApplicationModel;
+using Wox.Infrastructure.Storage;
+
+namespace Microsoft.Plugin.Program.Storage
+{
+    /// <summary>
+    /// A repository for storing packaged applications such as UWP apps or appx packaged desktop apps.
+    /// This repository will also monitor for changes to the PackageCatelog and update the repository accordingly
+    /// </summary>
+    internal class PackageRepository : ListRepository<UWP.Application>, IRepository<UWP.Application>, IProgramRepository
+    {
+        IPackageCatalog _packageCatalog;
+        public PackageRepository(IPackageCatalog packageCatalog, IStorage<IList<UWP.Application>> storage) : base(storage)
+        {
+            _packageCatalog = packageCatalog ?? throw new ArgumentNullException("packageCatalog", "PackageRepository expects an interface to be able to subscribe to package events");
+            _packageCatalog.PackageInstalling += OnPackageInstalling;
+            _packageCatalog.PackageUninstalling += OnPackageUninstalling;
+        }
+
+        public void OnPackageInstalling(PackageCatalog p, PackageInstallingEventArgs args)
+        {
+            if (args.IsComplete)
+            {
+
+                try
+                {
+                    var uwp = new UWP(args.Package);
+                    uwp.InitializeAppInfo(args.Package.InstalledLocation.Path);
+                    foreach (var app in uwp.Apps)
+                    {
+                        Add(app);
+                    }
+                }
+                //InitializeAppInfo will throw if there is no AppxManifest.xml for the package. 
+                //Note there are sometimes multiple packages per application and this doesn't necessarily mean that we haven't found the app.

```

Comment:
I think it is the other way round. We can have multiple apps in one package. Not sure if this would affect the logic of reindexing.

Ref - https://social.msdn.microsoft.com/forums/windowsapps/en-US/eef715ac-ba0e-41cc-8cfe-fd95051dae75/ipc-between-multiple-apps-in-same-package#:~:text=You%20can%20use%20the%20Applications%20element%20to%20specify,process.%203.%20Each%20Application%20gets%20its%20own%20tile..

Reply:
You are correct that there multiple apps per package, which is why there is a foreach in this method.    However this callback will get triggered multiple times for the same locical application. 

You can reproduce this by installing the terminal preview application from windows store.  OnPackageInstalling will get called multiple times one of which is in a location that doesn't have a appxmanifest.xaml file.  Note this is the consistent with when we index the method originally which is why the UWP.All method has a try catch around the folliwng lines as well: 

```
var uwp = new UWP(args.Package);
 uwp.InitializeAppInfo(args.Package.InstalledLocation.Path);
```

Reply:
Sounds good. [nit] can we change the comment accordingly? It says `there are sometimes multiple packages per application`.

Reply:
I gotcha. Yeah sure, any suggestion.  Multiple pacakges per `product`?

Reply:
Yea, that sounds good.